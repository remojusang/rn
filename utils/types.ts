export type RootStackParamList = {
  Tabs: undefined;
  Home: undefined;
  Login: { signedEmail: string } | undefined;
  Sign: undefined;
  Detail: PersonType;
  Camera: undefined;
  Profile: undefined;
};

export type PersonType = {
  name: string;
  age: number;
  gender: 'male' | 'female';
};

export type AlbumImageType = {
  uri: string;
};
