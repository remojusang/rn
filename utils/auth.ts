import auth from '@react-native-firebase/auth';

interface ISign {
  email: string;
  password: string;
}

function signIn({ email, password }: ISign) {
  return auth().signInWithEmailAndPassword(email, password);
}

function signUp({ email, password }: ISign) {
  return auth().createUserWithEmailAndPassword(email, password);
}

function subscribeAuth(callback: any) {
  return auth().onAuthStateChanged(callback);
}

function signOut() {
  return auth().signOut();
}

export { signIn, signUp, subscribeAuth, signOut };
