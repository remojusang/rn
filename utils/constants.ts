type RegexType = {
  email: RegExp;
  password: RegExp;
  phone: RegExp;
  passwordCheck?: RegExp; // TS에러 방지
};

const REGEX: RegexType = {
  email: /^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/,
  password: /^[a-zA-Z0-9]{8,16}$/,
  phone: /^\d{3}-\d{3,4}-\d{4}$/,
} as const;

type IndexStringObject = { [index: string]: string };

const FORM_ERR_MSG: IndexStringObject = {
  email: '올바른 이메일 패턴이 아닙니다.',
  password: '8~16자 영문 대소문자, 숫자를 사용하세요.',
  required: '해당란을 입력해주세요.',
  passwordCheck: '비밀번호가 일치하지 않습니다.',
  phone: '올바른 휴대폰 패턴이 아닙니다.',
} as const;

type accessHintType = {
  LOADER: 'loader';
  PW: 'password';
  PW_CHECK: 'password_check';
};

const ACCESS_HINT: accessHintType = {
  LOADER: 'loader',
  PW: 'password',
  PW_CHECK: 'password_check',
} as const;

const COUNTRIES = ['한글', 'English', '日本語'];

const countryToLocale: IndexStringObject = {
  한글: 'ko',
  English: 'en-US',
  日本語: 'jp',
} as const;

const CAMERA_ERRCODE: IndexStringObject = {
  camera_unavailable: 'camera not available on device',
  permission: 'Permission not satisfied',
} as const;

export {
  REGEX,
  FORM_ERR_MSG,
  ACCESS_HINT,
  COUNTRIES,
  countryToLocale,
  CAMERA_ERRCODE,
};
