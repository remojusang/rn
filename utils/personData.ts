import { PersonType } from './types';

const personData: PersonType[] = [
  {
    name: 'Kalid',
    age: 15,
    gender: 'male',
  },
  {
    name: 'Fomagran',
    age: 20,
    gender: 'male',
  },
  {
    name: 'May',
    age: 19,
    gender: 'female',
  },
  {
    name: 'Julia',
    age: 29,
    gender: 'female',
  },
  {
    name: 'Hunter',
    age: 18,
    gender: 'male',
  },
  {
    name: 'Jane',
    age: 28,
    gender: 'female',
  },
  {
    name: 'Kendrick',
    age: 34,
    gender: 'male',
  },
  {
    name: 'Kali',
    age: 25,
    gender: 'female',
  },
  {
    name: 'Jay',
    age: 37,
    gender: 'male',
  },
  {
    name: 'Mac',
    age: 31,
    gender: 'male',
  },
];

export default personData;
