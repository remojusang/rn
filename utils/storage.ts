import AsyncStorage from '@react-native-async-storage/async-storage';

const getStorageData = async (key: string) => {
  return JSON.parse((await AsyncStorage.getItem(key)) || 'null');
};

const setStorageData = async (key: string, value: any) => {
  await AsyncStorage.setItem(key, JSON.stringify(value));
};

const removeStorageData = async (key: string) =>
  await AsyncStorage.removeItem(key);

export { getStorageData, setStorageData, removeStorageData };
