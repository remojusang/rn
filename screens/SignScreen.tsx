import React, { useRef } from 'react';
import type { NativeStackScreenProps } from '@react-navigation/native-stack';
import { RootStackParamList } from '../utils/types';
import { Text, StyleSheet, TextInput, Alert } from 'react-native';
import { useForm } from 'react-hook-form';
import CustomBtn from '../components/CustomBtn';
import FormLayout from '../components/FormLayout';
import FormInput from '../components/FormInput';
import { ACCESS_HINT } from '../utils/constants';
import { useIntl } from 'react-intl';
import { useRecoilValue } from 'recoil';
import { localeState } from '../components/Atoms';
import FlexGap from '../components/FlexGap';

type SignProps = NativeStackScreenProps<RootStackParamList, 'Sign'>;

export interface ISignForm {
  email: string;
  password: string;
  passwordCheck: string;
  phone: string;
}
function SignScreen({ navigation }: SignProps) {
  const pwRef = useRef<TextInput>(null);
  const pwCheckRef = useRef<TextInput>(null);
  const phoneRef = useRef<TextInput>(null);
  const { formState, handleSubmit, control, getValues } =
    useForm<ISignForm>({ mode: 'onChange' });
  const onValid = ({ email, password, phone }: ISignForm) => {
    if (
      email !== 'test123@naver.com' ||
      password !== 'remo1234' ||
      phone !== '010-1234-5678'
    ) {
      return;
    }
    Alert.alert('알림', '회원가입 완료.', [
      {
        text: 'OK',
        onPress: () =>
          navigation.navigate('Login', { signedEmail: email }),
      },
    ]);
  };
  const { formatMessage } = useIntl();
  const locale = useRecoilValue(localeState);
  return (
    <FormLayout>
      <>
        <FormInput
          name="email"
          errorMsg={formState.errors.email?.message}
          control={control}
          textInputConfig={{
            keyboardType: 'email-address',
            placeholder: 'remo@naver.com',
          }}
          onNext={() => pwRef.current?.focus()}
        />
        <FormInput
          constraintslabel={formatMessage({
            id: 'passwordConstraintslabel',
          })}
          name="password"
          errorMsg={formState.errors.password?.message}
          control={control}
          textInputConfig={{
            placeholder: '********',
            secureTextEntry: true,
          }}
          accessibilityHint={ACCESS_HINT.PW}
          onNext={() => pwCheckRef.current?.focus()}
          inputRef={pwRef}
        />
        <FormInput
          name="passwordCheck"
          passwordVal={getValues('password')}
          errorMsg={formState.errors.passwordCheck?.message}
          control={control}
          textInputConfig={{
            placeholder: '********',
            secureTextEntry: true,
          }}
          accessibilityHint={ACCESS_HINT.PW_CHECK}
          onNext={() => phoneRef.current?.focus()}
          inputRef={pwCheckRef}
        />
        <FormInput
          constraintslabel={formatMessage({
            id: 'phoneConstraintslabel',
          })}
          name="phone"
          errorMsg={formState.errors.phone?.message}
          control={control}
          textInputConfig={{
            keyboardType: 'phone-pad',
            placeholder: '010-1234-5678',
          }}
          inputRef={phoneRef}
        />
        <FlexGap gapSize={10} />

        {locale === 'ko' && (
          <Text style={styles.tos}>
            아래의 회원가입 버튼을 누르면, {'\n'}{' '}
            <Text style={styles.strong}>이용규약</Text>과{' '}
            <Text style={styles.strong}>프라이버시 정책</Text>에
            동의한 것이 됩니다.
          </Text>
        )}
        {locale === 'en-US' && (
          <Text style={styles.tos}>
            By clicking the sign up button below, {'\n'} you agree to
            the <Text style={styles.strong}>Terms of Use</Text> and{' '}
            <Text style={styles.strong}>Privacy Policy</Text>
          </Text>
        )}
        {locale === 'jp' && (
          <Text style={styles.tos}>
            下の会員登録ボタンを押すと, {'\n'}{' '}
            <Text style={styles.strong}>利用規約</Text> と{' '}
            <Text style={styles.strong}>プライバシーポリシー</Text>
            に同意したことになります
          </Text>
        )}
        <CustomBtn
          isLoading={false}
          title="signUpBtn"
          onPress={handleSubmit(onValid)}
        />
      </>
    </FormLayout>
  );
}

export default SignScreen;

const styles = StyleSheet.create({
  // termOfService
  tos: {
    width: '100%',
    fontSize: 14,
    marginBottom: 20,
  },
  strong: {
    color: '#4583FF',
  },
});
