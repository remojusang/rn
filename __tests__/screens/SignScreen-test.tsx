import {
  screen,
  render,
  fireEvent,
  act,
} from '@testing-library/react-native';
import { FORM_ERR_MSG, ACCESS_HINT } from '../../utils/constants';
import SignScreen from '../../screens/SignScreen';
import '@testing-library/jest-native/extend-expect';
const ErrorBoundary = require('../../components/ErrorBoundary');
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper'); // 에러방지코드: https://stackoverflow.com/questions/59587799/how-to-resolve-animated-usenativedriver-is-not-supported-because-the-native

describe('SignScreen', () => {
  const TEST_PROPS: any = {
    route: jest.fn(),
    navigation: jest.fn(),
  };
  test('렌더링 테스트', () => {
    const renderedJson = render(
      <ErrorBoundary>
        <SignScreen {...TEST_PROPS} />,
      </ErrorBoundary>,
    ).toJSON();
    expect(renderedJson).toMatchSnapshot();
    expect(renderedJson).toBeTruthy();
  });

  test('패스워드확인 유효성 검사', async () => {
    render(
      <ErrorBoundary>
        <SignScreen {...TEST_PROPS} />
      </ErrorBoundary>,
    );

    await act(async () => {
      fireEvent.changeText(
        await screen.findByAccessibilityHint(ACCESS_HINT.PW),
        '1234',
      );

      fireEvent.changeText(
        await screen.findByAccessibilityHint(ACCESS_HINT.PW_CHECK),
        '7777',
      );
    });

    // 비밀번호를 다르게 입력한 경우 에러메세지 출력되어야 한다.
    expect(
      await screen.findByText(FORM_ERR_MSG.passwordCheck),
    ).toBeVisible();

    await act(async () => {
      fireEvent.changeText(
        await screen.findByAccessibilityHint(ACCESS_HINT.PW_CHECK),
        '1234',
      );
    });
    // 비밀번호를 같게 입력한 경우 에러메세지 unmount 되어야 한다.
    expect(
      await screen.queryByText(FORM_ERR_MSG.passwordCheck),
    ).toBeNull();
  });
});

afterEach(() => {
  jest.clearAllMocks();
});
