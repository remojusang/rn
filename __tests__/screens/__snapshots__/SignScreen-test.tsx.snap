// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignScreen 렌더링 테스트 1`] = `
<RCTSafeAreaView
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "margin": 20,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "bottom": 0,
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "marginVertical": 16,
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            이메일
          </Text>
          <Text
            style={
              Object {
                "fontWeight": "300",
              }
            }
          />
        </View>
        <TextInput
          accessibilityHint=""
          autoCapitalize="none"
          keyboardType="email-address"
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder="remo@naver.com"
          style={
            Object {
              "backgroundColor": "white",
              "borderColor": "#000",
              "borderRadius": 6,
              "borderWidth": 1,
              "fontSize": 18,
              "padding": 10,
            }
          }
        />
      </View>
      <View
        collapsable={false}
        style={
          Object {
            "marginVertical": 16,
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            비밀번호
          </Text>
          <Text
            style={
              Object {
                "fontWeight": "300",
              }
            }
          >
            8~16자의 영문 대소문자와 숫자 조합
          </Text>
        </View>
        <TextInput
          accessibilityHint="password"
          autoCapitalize="none"
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder="********"
          secureTextEntry={true}
          style={
            Object {
              "backgroundColor": "white",
              "borderColor": "#000",
              "borderRadius": 6,
              "borderWidth": 1,
              "fontSize": 18,
              "padding": 10,
            }
          }
        />
      </View>
      <View
        collapsable={false}
        style={
          Object {
            "marginVertical": 16,
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            비밀번호 확인
          </Text>
          <Text
            style={
              Object {
                "fontWeight": "300",
              }
            }
          />
        </View>
        <TextInput
          accessibilityHint="password_check"
          autoCapitalize="none"
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder="********"
          secureTextEntry={true}
          style={
            Object {
              "backgroundColor": "white",
              "borderColor": "#000",
              "borderRadius": 6,
              "borderWidth": 1,
              "fontSize": 18,
              "padding": 10,
            }
          }
        />
      </View>
      <View
        collapsable={false}
        style={
          Object {
            "marginVertical": 16,
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            전화번호
          </Text>
          <Text
            style={
              Object {
                "fontWeight": "300",
              }
            }
          >
            하이픈 포함한 숫자만 입력
          </Text>
        </View>
        <TextInput
          accessibilityHint=""
          autoCapitalize="none"
          keyboardType="phone-pad"
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder="010-1234-5678"
          style={
            Object {
              "backgroundColor": "white",
              "borderColor": "#000",
              "borderRadius": 6,
              "borderWidth": 1,
              "fontSize": 18,
              "padding": 10,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "marginVertical": 5,
          }
        }
      />
      <Text
        style={
          Object {
            "fontSize": 14,
            "marginBottom": 20,
            "width": "100%",
          }
        }
      >
        아래의 회원가입 버튼을 누르면, 
        

         
        <Text
          style={
            Object {
              "color": "#4583FF",
            }
          }
        >
          이용규약
        </Text>
        과
         
        <Text
          style={
            Object {
              "color": "#4583FF",
            }
          }
        >
          프라이버시 정책
        </Text>
        에 동의한 것이 됩니다.
      </Text>
      <View
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#2F2F33",
              "borderRadius": 10,
              "padding": 12,
            }
          }
        >
          <ActivityIndicator
            accessibilityHint="loader"
            color="white"
          />
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
